Replace use of old APIs which are deprecated in OpenSSL 3

diff --git a/src/OpenSSL/SSL.py b/src/OpenSSL/SSL.py
index e71b044..b265a9d 100644
--- a/src/OpenSSL/SSL.py
+++ b/src/OpenSSL/SSL.py
@@ -1438,9 +1438,11 @@ class Context(object):
             _raise_current_error()
         bio = _ffi.gc(bio, _lib.BIO_free)
 
-        dh = _lib.PEM_read_bio_DHparams(bio, _ffi.NULL, _ffi.NULL, _ffi.NULL)
-        dh = _ffi.gc(dh, _lib.DH_free)
-        res = _lib.SSL_CTX_set_tmp_dh(self._context, dh)
+        pkey = _lib.PEM_read_bio_Parameters(bio, _ffi.NULL)
+        _openssl_assert(pkey != _ffi.NULL)
+        res = _lib.SSL_CTX_set0_tmp_dh_pkey(self._context, pkey)
+        if res != 1:
+            _lib.EVP_PKEY_free(pkey)
         _openssl_assert(res == 1)
 
     @_require_not_used
@@ -1467,7 +1469,7 @@
                 DeprecationWarning,
                 stacklevel=2,
             )
-            _lib.SSL_CTX_set_tmp_ecdh(self._context, curve._to_EC_KEY())
+            nid = curve._nid
         else:
             name = curve.name
             if name == "secp192r1":
@@ -1477,11 +1479,7 @@
             nid = _lib.OBJ_txt2nid(name.encode())
             if nid == _lib.NID_undef:
                 _raise_current_error()
-
-            ec = _lib.EC_KEY_new_by_curve_name(nid)
-            _openssl_assert(ec != _ffi.NULL)
-            ec = _ffi.gc(ec, _lib.EC_KEY_free)
-            _lib.SSL_CTX_set_tmp_ecdh(self._context, ec)
+        _lib.SSL_CTX_set1_curves(self._context, _ffi.new("int *", nid), 1)
 
     @_require_not_used
     def set_cipher_list(self, cipher_list: bytes) -> None:
diff --git a/src/OpenSSL/_util.py b/src/OpenSSL/_util.py
index 53c0b9e..a2c81f1 100644
--- a/src/OpenSSL/_util.py
+++ b/src/OpenSSL/_util.py
@@ -54,7 +54,7 @@
         errors.append(
             (
                 text(lib.ERR_lib_error_string(error)),
-                text(lib.ERR_func_error_string(error)),
+                "",
                 text(lib.ERR_reason_error_string(error)),
             )
         )
--- a/src/OpenSSL/crypto.py
+++ b/src/OpenSSL/crypto.py
@@ -364,35 +364,19 @@
             if bits <= 0:
                 raise ValueError("Invalid number of bits")
 
-            # TODO Check error return
-            exponent = _lib.BN_new()
-            exponent = _ffi.gc(exponent, _lib.BN_free)
-            _lib.BN_set_word(exponent, _lib.RSA_F4)
-
-            rsa = _lib.RSA_new()
-
-            result = _lib.RSA_generate_key_ex(rsa, bits, exponent, _ffi.NULL)
-            _openssl_assert(result == 1)
-
-            result = _lib.EVP_PKEY_assign_RSA(self._pkey, rsa)
-            _openssl_assert(result == 1)
+            crypto_key = rsa.generate_private_key(_lib.RSA_F4, bits)
+            tmp_pkey = PKey.from_cryptography_key(crypto_key)
 
         elif type == TYPE_DSA:
-            dsa = _lib.DSA_new()
-            _openssl_assert(dsa != _ffi.NULL)
-
-            dsa = _ffi.gc(dsa, _lib.DSA_free)
-            res = _lib.DSA_generate_parameters_ex(
-                dsa, bits, _ffi.NULL, 0, _ffi.NULL, _ffi.NULL, _ffi.NULL
-            )
-            _openssl_assert(res == 1)
-
-            _openssl_assert(_lib.DSA_generate_key(dsa) == 1)
-            _openssl_assert(_lib.EVP_PKEY_set1_DSA(self._pkey, dsa) == 1)
+            crypto_key = dsa.generate_private_key(bits)
+            tmp_pkey = PKey.from_cryptography_key(crypto_key)
         else:
             raise Error("No such key type")
 
-        self._initialized = True
+        # Move pkey contents to this object
+        self._only_public = tmp_pkey._only_public
+        self._pkey, tmp_pkey._pkey = tmp_pkey._pkey, None
+        self._initialized = tmp_pkey._initialized
 
     def check(self) -> bool:
         """
@@ -413,9 +397,11 @@
         if _lib.EVP_PKEY_type(self.type()) != _lib.EVP_PKEY_RSA:
             raise TypeError("Only RSA keys can currently be checked.")
 
-        rsa = _lib.EVP_PKEY_get1_RSA(self._pkey)
-        rsa = _ffi.gc(rsa, _lib.RSA_free)
-        result = _lib.RSA_check_key(rsa)
+        ctx = _lib.EVP_PKEY_CTX_new(self._pkey, _ffi.NULL)
+        _openssl_assert(ctx != _ffi.NULL)
+        ctx = _ffi.gc(ctx, _lib.EVP_PKEY_CTX_free)
+
+        result = _lib.EVP_PKEY_check(ctx)
         if result == 1:
             return True
         _raise_current_error()
@@ -529,16 +515,6 @@
     def __repr__(self) -> str:
         return f"<Curve {self.name!r}>"
 
-    def _to_EC_KEY(self) -> Any:
-        """
-        Create a new OpenSSL EC_KEY structure initialized to use this curve.
-
-        The structure is automatically garbage collected when the Python object
-        is garbage collected.
-        """
-        key = self._lib.EC_KEY_new_by_curve_name(self._nid)
-        return _ffi.gc(key, _lib.EC_KEY_free)
-
 
 @deprecated(
     "get_elliptic_curves is deprecated. You should use the APIs in "
@@ -2189,11 +2165,18 @@
     elif type == FILETYPE_ASN1:
         result_code = _lib.i2d_PrivateKey_bio(bio, pkey._pkey)
     elif type == FILETYPE_TEXT:
-        if _lib.EVP_PKEY_id(pkey._pkey) != _lib.EVP_PKEY_RSA:
+        if _lib.EVP_PKEY_type(pkey.type()) != _lib.EVP_PKEY_RSA:
             raise TypeError("Only RSA keys are supported for FILETYPE_TEXT")
 
-        rsa = _ffi.gc(_lib.EVP_PKEY_get1_RSA(pkey._pkey), _lib.RSA_free)
-        result_code = _lib.RSA_print(bio, rsa, 0)
+        result_code = _lib.EVP_PKEY_print_params(bio, pkey._pkey, 0, _ffi.NULL)
+        if result_code == 1:
+            result_code = _lib.EVP_PKEY_print_public(
+                bio, pkey._pkey, 0, _ffi.NULL
+            )
+        if result_code == 1:
+            result_code =_lib.EVP_PKEY_print_private(
+                bio, pkey._pkey, 0, _ffi.NULL
+            )
     else:
         raise ValueError(
             "type argument must be FILETYPE_PEM, FILETYPE_ASN1, or "
--- a/tests/test_crypto.py
+++ b/tests/test_crypto.py
@@ -1268,10 +1268,10 @@
         `PKey.generate_key` generates a DSA key when passed `TYPE_DSA` as a
         type and a reasonable number of bits.
         """
-        # 512 is a magic number.  The DSS (Digital Signature Standard)
-        # allows a minimum of 512 bits for DSA.  DSA_generate_parameters
-        # will silently promote any value below 512 to 512.
-        bits = 512
+        # 1024 is a magic number.  The DSS (Digital Signature Standard)
+        # allows a minimum of 1024 bits for DSA.  DSA_generate_parameters
+        # will silently promote any value below 1024 to 1024.
+        bits = 1024
         key = PKey()
         key.generate_key(TYPE_DSA, bits)
         assert key.type() == TYPE_DSA
@@ -1285,7 +1285,7 @@
         generate new keys.
         """
         key = PKey()
-        for type, bits in [(TYPE_RSA, 2048), (TYPE_DSA, 576)]:
+        for type, bits in [(TYPE_RSA, 2048), (TYPE_DSA, 1024)]:
             key.generate_key(type, bits)
             assert key.type() == type
             assert key.bits() == bits
@@ -2699,7 +2699,7 @@
         not RSA.
         """
         key = PKey()
-        key.generate_key(TYPE_DSA, 512)
+        key.generate_key(TYPE_DSA, 1024)
         with pytest.raises(TypeError):
             dump_privatekey(FILETYPE_TEXT, key)
 
@@ -3537,6 +3537,7 @@
         curve = next(iter(curves))
         assert f"<Curve {curve.name!r}>" == repr(curve)
 
+    @pytest.mark.skip("EC_KEY not supported")
     def test_to_EC_KEY(self) -> None:
         """
         The curve object can export a version of itself as an EC_KEY* via the
